src/Token/TokenLabel.tsx

```diff
@@ -0,0 +1,178 @@
+import React, {forwardRef, MouseEventHandler} from 'react'
+import styled, {css} from 'styled-components'
+import TokenBase, {isTokenInteractive, TokenBaseProps} from './TokenBase'
+import RemoveTokenButton from './_RemoveTokenButton'
+import tinycolor from 'tinycolor2'
+import {useTheme} from '../ThemeProvider'
+
+interface ColorModeConfig {
+  bgOpacity: number
+  borderThreshold: number
+  borderOpacity: number
+  lightnessThreshold: number
+}
+
+const colorModeConfigs: Record<string, ColorModeConfig> = {
+  dark: {
+    bgOpacity: 0.18,
+    borderThreshold: 0,
+    borderOpacity: 0.3,
+    lightnessThreshold: 0.6
+  },
+  light: {
+    bgOpacity: 1,
+    borderThreshold: 0.96,
+    borderOpacity: 1,
+    lightnessThreshold: 0.453
+  }
+}
+
+export interface TokenLabelProps extends TokenBaseProps {
+  /**
+   * The color that corresponds to the label
+   */
+  fillColor?: string
+  /**
+   * Whether the remove button should be rendered in the token
+   */
+  hideRemoveButton?: boolean
+}
+
+interface LabelStyleProps {
+  bgColor: React.CSSProperties['backgroundColor']
+  borderColor: React.CSSProperties['borderColor']
+  textColor: React.CSSProperties['color']
+}
+
+const tokenBorderWidthPx = 1
+
+const StyledTokenLabel = styled(TokenBase)<TokenLabelProps & LabelStyleProps>`
+  background-color: ${props => props.bgColor};
+  border-width: ${tokenBorderWidthPx}px;
+  border-style: solid;
+  border-color: ${props => props.borderColor};
+  color: ${props => props.textColor};
+  padding-right: ${props => (!props.hideRemoveButton ? 0 : undefined)};
+  position: relative;
+
+  ${props => {
+    if (props.isSelected) {
+      return css`
+        &:after {
+          content: '';
+          position: absolute;
+          z-index: 1;
+          top: -2px;
+          right: -2px;
+          bottom: -2px;
+          left: -2px;
+          display: block;
+          pointer-events: none;
+          box-shadow: 0 0 0 2px ${props.bgColor};
+          border-radius: 999px;
+        }
+      `
+    }
+  }}
+`
+
+const TokenTextContainer = styled('span')`
+  flex-grow: 1;
+  min-width: 0;
+  overflow: hidden;
+  text-overflow: ellipsis;
+  white-space: nowrap;
+`
+
+const TokenLabel = forwardRef<HTMLElement, TokenLabelProps>((props, forwardedRef) => {
+  const {as, fillColor, handleRemove, id, isSelected, ref, text, variant, hideRemoveButton, ...rest} = props
+  const {colorScheme} = useTheme()
+  const {bgOpacity, borderOpacity, borderThreshold, lightnessThreshold} = colorModeConfigs[colorScheme || 'light']
+  let bgColor = fillColor
+  let borderColor = fillColor
+  let textColor = '#FFF'
+  const perceivedLightness = tinycolor(fillColor).getLuminance()
+  const isFillColorLight = perceivedLightness >= lightnessThreshold
```

Comment:
Have you seen the Memex/dotcom implementation of issue labels? Maybe we can copy some of that code for LabelToken.

https://github.com/github/memex/blob/main/src/client/components/issue-label.tsx

Reply:
I saw the dotcom implementation, but not the Memex implementation. I'm definitely going to copy some of that code.

Reply:
I tried copying that code, but it's leading to bad contrast, I can't figure out how to get the `:after` pseudo-element styles working, and it's going to take a little more work to get the `darken` and `lighten` color functions right for the `isSelected` styles.

![Screen Shot 2021-10-04 at 8 29 50 PM](https://user-images.githubusercontent.com/2313998/135942141-ac0653a9-ed80-4d86-92c1-4547e4bd81d5.png)

If you still think we should still use the custom property implementation, I'll make it work. Otherwise, I'm just going to replace `tinycolor2` with `color2k` and keep everything in JavaScript.

Reply:
Hmm looks like this is consistent with how dotcom renders labels. Guess it does technically pass color contrast 
<img width="359" alt="CleanShot 2021-10-05 at 13 43 33@2x" src="https://user-images.githubusercontent.com/4608155/136099790-1abaeb23-a319-463c-b5ea-9086d3f96705.png">

Reply:
I just went with the same implementation we have for Memex